quickwit/quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index/mod.rs

```diff
@@ -165,20 +165,15 @@ impl FileBackedIndex {
         &self.splits
     }
 
-    pub(crate) fn stage_split(
-        &mut self,
-        split_metadata: SplitMetadata,
-    ) -> crate::MetastoreResult<()> {
+    /// Stages a single split returning if the split was inserted or not.
+    pub(crate) fn stage_split(&mut self, split_metadata: SplitMetadata) -> bool {
         // Check whether the split exists.
-        // If the split exists, return an error to prevent the split from being registered.
-        if self.splits.contains_key(split_metadata.split_id()) {
-            return Err(MetastoreError::InternalError {
-                message: format!(
-                    "Failed to stage split  `{}`: split already exists.",
-                    split_metadata.split_id()
-                ),
-                cause: "".to_string(),
-            });
+        // If the split exists, we check what state it is in. If it's anything other than `Staged`
+        // something has gone very wrong and we should abort the operation.
+        if let Some(existing_split) = self.splits.get(split_metadata.split_id()) {
+            if existing_split.split_state != SplitState::Staged {
+                return false;
```

Comment:
The contract in the method doc is imprecise.
If anything it does into seem to describe the current implementation.

It suggest the return value is similar to a return bool of `Set::add`.
If the value was already there `return false`, if it was not `return true`.

In other methods, this is also what it means, it is used to identify whether it is necessary or not to write the file.

In your new implem, it is used to mark a failure, which seems like a poor choice.

Reply:
Switched to returning and matching on a `MetastoreError`, docs have been updated.